File: graph/Keys and Rooms/Keys and Rooms.cpp

class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        vector<bool> visit(rooms.size(), false);
        visit[0] = true;
        int i = 0;
        queue<int> q;
        while( i < rooms.size() ) {
            for(int j = 0; j < rooms[i].size();j++){
                if(visit[rooms[i][j]] == false){
                   q.push(rooms[i][j]);
                     visit[rooms[i][j]] = true;
                }
            }
            i = q.front();
            q.pop();
        }
        for(int i = 0; i < visit.size(); i++)
            if(visit[i] == false) return false;
        return true;
    }
};

File: graph/Keys and Rooms/Keys and Rooms.java

class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        ArrayDeque<Integer> que = new ArrayDeque<>();
        que.push(0);
        HashSet<Integer> set = new HashSet<>();
        int n = rooms.size();
        boolean[] visited = new boolean[n];
        while(que.size()>0){
            int rem = que.removeFirst();
            visited[rem] = true;
            set.add(rem);
            for(int k:rooms.get(rem)){
                if(visited[k]==false)
                    que.push(k);
            }
        }
        return set.size()==n;
    }
}
